nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/NiFiAtlasClient.java

```diff
@@ -0,0 +1,537 @@
+     * If Atlas is configured to use soft-delete (default), deleted ids are still returned.
+     * Fetched entities are used to determine whether an AtlasObjectId is still active or deleted.
+     * Deleted entities will not be included in the result of this method.
+     * </p>
+     * @param ids to convert
+     * @return AtlasObjectIds with qualifiedName
+     */
+    private Map<AtlasObjectId, AtlasEntity> toQualifiedNameIds(List<AtlasObjectId> ids) {
+        if (ids == null) {
+            return Collections.emptyMap();
+        }
+
+        return ids.stream().distinct().map(id -> {
+            try {
+                final AtlasEntity.AtlasEntityWithExtInfo entityExt = searchEntityDef(id);
+                final AtlasEntity entity = entityExt.getEntity();
+                if (AtlasEntity.Status.DELETED.equals(entity.getStatus())) {
+                    return null;
+                }
+                final Map<String, Object> uniqueAttrs = Collections.singletonMap(ATTR_QUALIFIED_NAME, entity.getAttribute(ATTR_QUALIFIED_NAME));
+                return new Tuple<>(new AtlasObjectId(id.getGuid(), id.getTypeName(), uniqueAttrs), entity);
+            } catch (AtlasServiceException e) {
+                return null;
+            }
+        }).filter(Objects::nonNull).collect(Collectors.toMap(Tuple::getKey, Tuple::getValue));
+    }
+
+    public void registerNiFiFlow(NiFiFlow nifiFlow) throws AtlasServiceException {
+
+        // Create parent flow entity, so that common properties are taken over.
+        final AtlasEntity flowEntity = registerNiFiFlowEntity(nifiFlow);
+
+        // Create DataSet entities those are created by this NiFi flow.
+        final Map<String, List<AtlasEntity>> updatedDataSetEntities = registerDataSetEntities(nifiFlow);
+
+        // Create path entities.
+        final Set<AtlasObjectId> remainingPathIds = registerFlowPathEntities(nifiFlow);
+
+        // Update these attributes only if anything is created, updated or removed.
+        boolean shouldUpdateNiFiFlow = nifiFlow.isMetadataUpdated();
+        if (remainingPathIds != null) {
+            flowEntity.setAttribute(ATTR_FLOW_PATHS, remainingPathIds);
+            shouldUpdateNiFiFlow = true;
+        }
+        if (updatedDataSetEntities.containsKey(TYPE_NIFI_QUEUE)) {
+            flowEntity.setAttribute(ATTR_QUEUES, updatedDataSetEntities.get(TYPE_NIFI_QUEUE));
+            shouldUpdateNiFiFlow = true;
+        }
+        if (updatedDataSetEntities.containsKey(TYPE_NIFI_INPUT_PORT)) {
+            flowEntity.setAttribute(ATTR_INPUT_PORTS, updatedDataSetEntities.get(TYPE_NIFI_INPUT_PORT));
+            shouldUpdateNiFiFlow = true;
+        }
+        if (updatedDataSetEntities.containsKey(TYPE_NIFI_OUTPUT_PORT)) {
+            flowEntity.setAttribute(ATTR_OUTPUT_PORTS, updatedDataSetEntities.get(TYPE_NIFI_OUTPUT_PORT));
+            shouldUpdateNiFiFlow = true;
+        }
+
+        logger.debug("### NiFi Flow Audit Logs START");
```

Comment:
This whole block should probably be wrapped in a check if logger.isDebugEnabled()

Reply:
Indeed. Thanks for the advice!